File: LeetCodeExample.Test/DP/_00139_WordBreak.py
# Given a string s and a dictionary of strings wordDict, return true if s can be segmented into a space-separated sequence of one or more dictionary words.

# Note that the same word in the dictionary may be reused multiple times in the segmentation.

from typing import List
from collections import defaultdict
class Trie:
    def __init__(self):
        self.childs = defaultdict(Trie)
        self.isWord = False

    # def add(word: str):


class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        root = Trie()
        def buildTrie():
            # build Trie
            for word in wordDict:
                node = root
                for c in word:
                    # defaultdict takes care of the "not exists" condition
                    node = node.childs[c]
                node.isWord = True

        # def searchTrie()

        buildTrie()
        # Record if we can finish a word sequence from the given index
        dp = [False] * len(s)
        # Start from the end
        for i in range(len(s) - 1, -1, -1):
            node = root
            j = i
            while j < len(s) and node:
                if s[j] not in node.childs:
                    break
                node = node.childs[s[j]]
                if node.isWord:
                    # Is this the last word? Or, is the next word a continuation to the last word?
                    if j == len(s) - 1 or dp[j + 1]:
                        dp[i] = True
                        break
                # We continue down the Trie if it is not yet a word
                j += 1
        return dp[0]

File: LeetCodeExample.Test/DP/_00139_WordBreak.cs
using NUnit.Framework;
using System;
using System.Collections.Generic;
using System.Linq;

namespace LeetCodeExample.Test
{
    /// <summary>
    // Given a string s and a dictionary of strings wordDict, return true if s can be segmented into a space-separated sequence of one or more dictionary words.
    // Note that the same word in the dictionary may be reused multiple times in the segmentation.
    /// </summary>
    public class _00139_WordBreak
    {
        // Would be faster to store the dictionary in a Trie, change to a Trie when I have time
        //HashSet<string> Dict = new HashSet<string>();
        class Node
        {
            public char ch;
            public bool isWord = false;
            public Dictionary<char, Node> children = new Dictionary<char, Node>();
        }

        Node Trie = new Node();

        public bool WordBreak(string s, IList<string> wordDict)
        {
            foreach (var word in wordDict)
            {
                Node trie = Trie;
                foreach (char c in word)
                {
                    if (!trie.children.ContainsKey(c))
                        trie.children.Add(c, new Node());
                    trie = trie.children[c];
                }
                trie.isWord = true;
                //Dict.Add(word);
            }

            return dp(s, 0);
        }

        Dictionary<int, bool> memo = new Dictionary<int, bool>();

        bool dp(string s, int start)
        {
            // base case
            if (start >= s.Length)
                return true;

            if (memo.ContainsKey(start))
                return memo[start];

            // Max size of a word in the dictionary is 20 characters
            //string sub = "";
            var trie = Trie;
            for (int i = 0; i < 20 && i + start < s.Length; i++)
            {
                char ch = s[start + i];
                if (!trie.children.ContainsKey(ch))
                    break;
                trie = trie.children[ch];
                if (!trie.isWord)
                    continue;

                //string sub = s.Substring(start, i);
                //sub += s[start + i - 1];
                //Console.WriteLine($"Sub: {sub}");
                //if (!Dict.Contains(sub))
                //    continue;
                bool res = dp(s, start + i + 1);
                if (res)
                {
                    memo.Add(start, true);
                    return true;
                }
            }
            memo.Add(start, false);
            return false;
        }
    }
}